lib/ansible/modules/cloud/amazon/rds_instance.py

```diff
@@ -0,0 +1,993 @@
+
+    # If we're not waiting for a delete to complete then we're all done
+    # so just return
+    if not module.params.get('wait'):
+        return dict(changed=True, response=response)
+    try:
+        instance = await_resource(conn, instance_id, 'deleted', module)
+    except botocore.exceptions.ClientError as e:
+        if e.code == 'DBInstanceNotFound':
+            return dict(changed=True)
+        else:
+            module.fail_json_aws(e, msg="waiting for rds deletion to complete")
+    except Exception as e:
+            module.fail_json_aws(e, msg="waiting for rds deletion to complete")
+
+    return dict(changed=True, response=response, instance=instance)
+
+
+def update_rds_tags(module, conn, db_instance=None):
+    # if we get no db_instance we should go collect one; not needed
+    # now - we currently inherit the one modify ends up with
+    assert db_instance is not None, "internal error - no db_instance in update_rds_tags call"
+
+    db_instance_arn = db_instance['DBInstanceArn']
+
+    # it's much easier to do set manipulation in ansible form
+    current_tag_list = conn.list_tags_for_resource(db_instance_arn)['TagList']
+    current_tags = boto3_tag_list_to_ansible_dict(current_tag_list)
+    desired_tags = module.params.get('tags')
+
+    if desired_tags is None:
+        return False
+
+    changed = False
+
+    current_keys = set(current_tags.keys())
+    desired_keys = set(desired_tags.keys())
+    to_add = desired_keys - current_keys
+    to_delete = current_keys - desired_keys
+
+    for k in desired_keys & current_keys:
+        if current_tags[k] != desired_tags[k]:
+            to_delete.add(k)
+            to_add.add(k)
+
+    if to_delete:
+        try:
+            conn.remove_tags_from_resource(ResourceName=db_instance_arn, TagKeys=list(to_delete))
+        except botocore.exceptions.ClientError as e:
+            module.fail_json_aws(e, msg="trying to remove old tags from instance")
+        changed = True
+
+    if to_add:
+        tag_map = [{'Key': k, 'Value': desired_tags[k]} for k in to_add]
+        try:
+            conn.add_tags_to_resource(ResourceName=db_instance_arn, Tags=tag_map)
+        except botocore.exceptions.ClientError as e:
+            module.fail_json_aws(e, msg="trying to add new tags to instance")
+        changed = True
+
+    return changed
+
+
+def abort_on_impossible_changes(module, before_facts):
+    for immutable_key in ['username', 'engine', 'db_name']:
+        if immutable_key in module.params:
+            try:
+                keys_different = module.params[immutable_key] != before_facts[immutable_key]
+            except KeyError:
+                keys_different = False
+            if (keys_different):
+                module.fail_json(msg="Cannot modify parameter %s for instance %s" %
+                                 (immutable_key, before_facts['db_instance_identifier']))
+
+
+def snake_dict_to_cap_camel_dict(snake_dict):
```

Comment:
There are various versions lying around I think but the ones I found do pascalCase rather than CamelCase (I tried looking for one I could reuse).  Right now the function I have doesn't cover abbreviations properly either (as demonstrated by the bug you found).  

I think once we have a nice version then it should be merged to utils, however I'm developing it in a module until I'm confident it's sensible.

Reply:
See #31400